transformer_engine/pytorch/csrc/extensions/permutation.cu

```diff
@@ -0,0 +1,245 @@
+
+    size_t temp_storage_bytes = 0;
+    int *temp_ptr = nullptr;
+    cub::DeviceRadixSort::SortPairs(nullptr, temp_storage_bytes, temp_ptr, temp_ptr, temp_ptr,
+                                    temp_ptr, max_expanded_token_num);
+    Tensor temp_storage = torch::empty(
+        temp_storage_bytes, torch::dtype(torch::kInt8).device(torch::kCUDA).requires_grad(false));
+
+    workspace.push_back(sorted_indices);
+    workspace.push_back(row_id);
+    workspace.push_back(sorted_row_id);
+    workspace.push_back(temp_storage);
+  }
+
+  int *indices_ptr = reinterpret_cast<int *>(getDataPtr(indices, 0));
+  int *sorted_indices_ptr = reinterpret_cast<int *>(getDataPtr(workspace[0], 0));
+  int *row_id_ptr = reinterpret_cast<int *>(getDataPtr(workspace[1], 0));
+  int *sorted_row_id_ptr = reinterpret_cast<int *>(getDataPtr(workspace[2], 0));
+
+  void *d_temp_storage = getDataPtr(workspace[3], 0);
+  size_t temp_storage_bytes = std::numeric_limits<size_t>::max();
+
+  cub::DeviceRadixSort::SortPairs(d_temp_storage, temp_storage_bytes, indices_ptr,
+                                  sorted_indices_ptr, row_id_ptr, sorted_row_id_ptr,
+                                  num_tokens * num_topK);
+
+  // activations type
+  const at::ScalarType _st = input.scalar_type();
+
+  // Output buffer alloc
+  num_out_tokens = (num_out_tokens > 0) ? num_out_tokens : num_tokens * num_topK;
+  Tensor permuted_output = torch::empty(
+      {num_out_tokens, num_cols}, torch::dtype(_st).device(torch::kCUDA).requires_grad(false));
+  Tensor row_id_map =
+      torch::empty({num_tokens * num_topK},
+                   torch::dtype(torch::kInt32).device(torch::kCUDA).requires_grad(false));
+
+  int *row_id_map_ptr = reinterpret_cast<int *>(getDataPtr(row_id_map, 0));
+  auto stream = at::cuda::getCurrentCUDAStream().stream();
+
+  void *input_ptr = getDataPtr(input, 0);
+  void *permuted_output_ptr = getDataPtr(permuted_output, 0);
+
+  switch (_st) {
+    case at::ScalarType::Float: {
+      moe_permutation_launcher<float, true>(input_ptr, permuted_output_ptr, sorted_row_id_ptr,
+                                            row_id_map_ptr, nullptr, num_tokens, num_topK, num_cols,
+                                            num_out_tokens, stream);
+
+      break;
+    }
+    case at::ScalarType::Half: {
+      moe_permutation_launcher<half, true>(input_ptr, permuted_output_ptr, sorted_row_id_ptr,
+                                           row_id_map_ptr, nullptr, num_tokens, num_topK, num_cols,
+                                           num_out_tokens, stream);
+
+      break;
+    }
+    case at::ScalarType::BFloat16: {
+      moe_permutation_launcher<__nv_bfloat16, true>(
+          input_ptr, permuted_output_ptr, sorted_row_id_ptr, row_id_map_ptr, nullptr, num_tokens,
+          num_topK, num_cols, num_out_tokens, stream);
+
+      break;
+    }
+    case at::ScalarType::Float8_e5m2: {
```

Comment:
In TE, we are not using PyTorch's `float8_e5m2` and `float8_e4m3fn` data types. Instead, we just use `unit8` or `byte` tensors with an additional dtype indicating the real FP8 dtype.

Take `cast_from_fp8` for example:
Python code:
https://github.com/NVIDIA/TransformerEngine/blob/56e0b351d0b7db6e622d3aa3eac6c6a1bf1ce4ab/transformer_engine/pytorch/cpp_extensions/cast.py#L46-L60
C++ code: https://github.com/NVIDIA/TransformerEngine/blob/56e0b351d0b7db6e622d3aa3eac6c6a1bf1ce4ab/transformer_engine/pytorch/csrc/extensions/cast.cu#L43-L58

Reply:
Done.